From 3ef74baae5ba410d3b75bad0c204e22ee530416a Mon Sep 17 00:00:00 2001
From: "mujiburrahimank@maxlinear.com"
 <mujiburrahimank@lvdocker7.lvsw.maxlinear.com>
Date: Sun, 15 Feb 2026 23:32:12 -0800
Subject: [PATCH] WLANRTSYS-100494 : [WNC][LR72][CAS-41983-Q9Y8F2] AFC failed
 in BV certification lab -

Issue : hostapd_cli send_afc_request fail with AFCDUSA32 item and wlan4 interace was gone

RCA : There is no standard implementation to handle 6G AFC power mode
fallback to LPI mechanism

Fix : Implemented the new design which checks for the afc data
before switching mode and falls back to LPI if no valid data found in SP
power mode.

FIXUP to WLANRTSYS-83162 Ported afcd and composite mode changes from legacy hostapd
---
 hostapd/mxl_ctrl_iface.c        | 21 ++++++-----
 src/ap/afc.c                    | 38 +++++++++++++++++++
 src/ap/hostapd.h                |  3 ++
 src/ap/mxl_acs.h                |  1 +
 src/ap/mxl_acs_proprietary.c    | 67 +++++++++++++++++++++++++++++++--
 src/ap/mxl_acs_proprietary.h    |  2 +-
 src/ap/mxl_afc.c                | 60 ++++++++++++++++++++++++++++-
 src/ap/mxl_afc.h                |  4 +-
 src/ap/mxl_common_proprietary.c |  4 +-
 src/ap/mxl_dfs_common.c         |  2 +-
 src/ap/mxl_dfs_common.h         |  2 +-
 src/ap/mxl_events.c             | 12 +++++-
 src/ap/mxl_hostapd.c            |  4 ++
 src/ap/mxl_zwdfs.c              |  2 +-
 14 files changed, 199 insertions(+), 23 deletions(-)

diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index c4e7e0bae..591acd94a 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -89,6 +89,8 @@ static int mxl_hostapd_ctrl_iface_send_afcd_request(struct hostapd_iface *iface,
 					     char *buf, size_t buflen)
 {
 	mxl_hostapd_afc_delete_data_from_server(iface);
+	iface->afc.retry_count = 0;
+	iface->afc.he_6ghz_reg_pwr_type = iface->conf->he_6ghz_reg_pwr_type;
 	return mxl_hostapd_afc_handle_request(iface);
 }
 #endif /* CONFIG_AFC */
@@ -3943,9 +3945,16 @@ int mxl_hostapd_ctrl_iface_set_power_mode(struct hostapd_iface *iface, const cha
 	val = atoi(pos);
 	switch (val) {
 	case HE_REG_INFO_6GHZ_AP_TYPE_INDOOR:
+		wpa_printf(MSG_INFO, "LPI Power Mode is supported");
+		iface->conf->he_6ghz_reg_pwr_type = val;
+		mxl_hostapd_interface_update_power_regd(hapd, val);
+		break;
 	case HE_REG_INFO_6GHZ_AP_TYPE_SP:
 	case HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP:
-		iface->conf->he_6ghz_reg_pwr_type = val;
+		wpa_printf(MSG_INFO, "INDOOR_SP/SP Power Mode is supported");
+#ifdef CONFIG_AFC
+		mxl_hostapd_afc_get_valid_rule_and_set_power(iface, val);
+#endif /* CONFIG_AFC */
 		break;
 	case HE_REG_INFO_6GHZ_AP_TYPE_VLP:
 		wpa_printf(MSG_INFO, "VERY-LOW-POWER is not supported"); /* Fall Through */
@@ -3953,22 +3962,14 @@ int mxl_hostapd_ctrl_iface_set_power_mode(struct hostapd_iface *iface, const cha
 		wpa_printf(MSG_INFO, "Invalid power mode");
 		return -1;
 	}
-	mxl_hostapd_interface_update_power_regd(hapd, val);
-#ifdef CONFIG_AFC
-	if (val == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP) {
-		/* Apply UNII 5/7 LPI only if moving to Indoor SP */
-		mxl_hostapd_afc_clear_and_load_lpi(iface);
-		mxl_hostap_afc_construct_regrule_from_afc_data(iface);
-	}
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-	else if (val == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR) {
+	if (val == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR) {
 		if (iface->conf->mxl_conf.acs_6g_punct_mode) {
 			mxl_hostapd_set_eht_punct_bitmap(iface->bss[0], 0);
 			iface->conf->punct_bitmap = 0;
 		}
 	}
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
-#endif /* CONFIG_AFC */
 
 	return os_snprintf(buf, buflen, "Power Mode=%s\n", mxl_he_6ghz_pwr_mode_text(val));
 }
diff --git a/src/ap/afc.c b/src/ap/afc.c
index 50e81b14d..d5779156f 100644
--- a/src/ap/afc.c
+++ b/src/ap/afc.c
@@ -989,6 +989,37 @@ static void hostapd_afc_delete_data_from_server(struct hostapd_iface *iface)
 	iface->afc.data_valid = false;
 }
 
+static void hostapd_afc_lpi_timeout_handler(void *eloop_ctx, void *timeout_ctx)
+{
+	struct hostapd_iface *iface = eloop_ctx;
+	int ret = 0;
+
+	if (iface->afc.retry_count <= 5) {
+		ret = mxl_hostapd_afc_handle_request(iface);
+		if (ret < 0) {
+			wpa_printf(MSG_ERROR, "AFC request retry count %d", iface->afc.retry_count);
+			return;
+		}
+		ret = mxl_hostapd_interface_update_power_regd(iface->bss[0], iface->afc.he_6ghz_reg_pwr_type);
+	}
+	else {
+		if (iface->afc.he_6ghz_reg_pwr_type == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP) {
+#ifdef CONFIG_AFC
+			/* Apply UNII 5/7 LPI only if moving to Indoor SP */
+			iface->conf->he_6ghz_reg_pwr_type = iface->afc.he_6ghz_reg_pwr_type;
+			mxl_hostapd_afc_clear_and_load_lpi(iface);
+			mxl_hostap_afc_construct_regrule_from_afc_data(iface);
+			hostapd_set_state(iface, HAPD_IFACE_ACS);
+			iface->afc.acs_pwr_mode_6g = 1;
+			mxl_hostapd_interface_update_power_regd(iface->bss[0], iface->afc.he_6ghz_reg_pwr_type);
+#endif /* CONFIG_AFC */
+		}
+		else {
+			wpa_printf(MSG_ERROR, "AFC request failed staying in LPI mode");
+			return;
+		}
+	}
+}
 
 static void hostapd_afc_timeout_handler(void *eloop_ctx, void *timeout_ctx)
 {
@@ -1093,6 +1124,13 @@ void hostap_afc_disable_channels(struct hostapd_iface *iface)
 int mxl_hostapd_afc_send_receive(struct hostapd_iface *iface) { return hostapd_afc_send_receive(iface); }
 void mxl_hostapd_afc_delete_data_from_server(struct hostapd_iface *iface) { hostapd_afc_delete_data_from_server(iface); }
 int mxl_hostapd_afc_parse_reply(struct hostapd_iface *iface, char *reply) { return hostapd_afc_parse_reply(iface, reply); }
+void mxl_hostapd_afc_resched_lpi(void *eloop_data, unsigned int secs)
+{
+	eloop_cancel_timeout(hostapd_afc_lpi_timeout_handler, eloop_data, NULL);
+	eloop_register_timeout(secs, 0,
+					 hostapd_afc_lpi_timeout_handler, eloop_data, NULL);
+}
+
 void mxl_hostapd_afc_resched_timeout(void *eloop_data, unsigned int secs)
 {
 	eloop_cancel_timeout(hostapd_afc_timeout_handler, eloop_data, NULL);
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 99e322897..c259411cb 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -829,6 +829,9 @@ struct hostapd_iface {
 #ifdef CONFIG_AFC
 	struct {
 		int timeout;
+		int retry_count;
+		u8 he_6ghz_reg_pwr_type;
+		int acs_pwr_mode_6g;
 		unsigned int num_freq_range;
 		struct afc_freq_range_elem {
 			int low_freq;
diff --git a/src/ap/mxl_acs.h b/src/ap/mxl_acs.h
index e9d591606..753edf470 100644
--- a/src/ap/mxl_acs.h
+++ b/src/ap/mxl_acs.h
@@ -34,6 +34,7 @@ enum hostapd_chan_status mxl_hostapd_check_chans(struct hostapd_iface *iface);
 int hostapd_acs_completed_wrap(struct hostapd_iface *iface, int err);
 void mxl_acs_stop_bgscan_timer(struct hostapd_data *hapd);
 void mxl_acs_set_bgscan_interval(struct hostapd_data *hapd);
+int mxl_acs_switch_to_unii_5_7_channel(struct hostapd_iface *iface);
 
 #endif /* MXL_ACS_H */
 
diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 2170eb9ff..d695867ed 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -3749,7 +3749,11 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 
 	}
 
+#ifdef CONFIG_AFC
+	if (!iface->mxl_iface.sec_best_ch.flag && !iface->afc.acs_pwr_mode_6g) {
+#else /* CONFIG_AFC */
 	if (!iface->mxl_iface.sec_best_ch.flag) {
+#endif /* CONFIG_AFC */
 		if ((!iface->conf->mxl_conf.acs_bg_scan_do_switch && (SWR_BG_SCAN == switch_reason)) ||
 		    iface->mxl_iface.cac_required || !(iface->conf->mxl_conf.acs_init_done)) {
 			wpa_printf(MSG_DEBUG, "ACS: stay on channel reason SWR_BG_SCAN switch not enabled or cac required or acs init not done");
@@ -4376,7 +4380,7 @@ void hostapd_get_op_class(int chan, int width, int *op_class)
 
 static void
 acs_get_failsafe_indexes_per_bw(struct hostapd_iface *iface,
-				struct acs_failsafe_table *table, bool skip_usable)
+				struct acs_failsafe_table *table, bool skip_usable, bool only_unii_5_7)
 {
 	int i;
 	int freq;
@@ -4403,6 +4407,13 @@ acs_get_failsafe_indexes_per_bw(struct hostapd_iface *iface,
 			}
 		}
 
+		/* Skipping the channels that are not UNII-5 or UNII-7 */
+		if (is_6ghz_op_class(iface->conf->op_class) && only_unii_5_7) {
+			if(!(mxl_is_6ghz_unii5_unii7(candidate[i].freq))) {
+				continue;
+			}
+		}
+
 		/* Skipping the channels which are restricted or not in chan_list */
 		if (candidate[i].exclude_reason == ACS_EXCLUDE_DISABLED)
 			continue;
@@ -4466,7 +4477,7 @@ acs_get_failsafe_indexes_per_bw(struct hostapd_iface *iface,
 }
 
 struct acs_candidate_table *
-mxl_acs_get_failsafe_channel(struct hostapd_iface *iface, bool skip_usable)
+mxl_acs_get_failsafe_channel(struct hostapd_iface *iface, bool skip_usable, bool only_unii_5_7)
 {
 	int i;
 	int best_rank_idx = -1;
@@ -4484,7 +4495,7 @@ mxl_acs_get_failsafe_channel(struct hostapd_iface *iface, bool skip_usable)
 		return NULL;
 	}
 
-	acs_get_failsafe_indexes_per_bw(iface, &table, skip_usable);
+	acs_get_failsafe_indexes_per_bw(iface, &table, skip_usable, only_unii_5_7);
 
 	current_bw_mhz = mxl_hostapd_get_num_width(hostapd_get_oper_chwidth(iface->conf), iface->conf->secondary_channel);
 
@@ -4605,7 +4616,7 @@ bool mxl_get_second_best_channel_from_acs_table(struct hostapd_iface *iface)
 
 	iface->mxl_iface.sec_best_ch.flag = true;
 	os_memset(&settings, 0, sizeof(settings));
-	channel = mxl_dfs_get_channel_dynamic_failsafe(iface, &secondary_channel, &vht_oper_centr_freq_seg0_idx, &vht_oper_centr_freq_seg1_idx, false, &settings.freq_params);
+	channel = mxl_dfs_get_channel_dynamic_failsafe(iface, &secondary_channel, &vht_oper_centr_freq_seg0_idx, &vht_oper_centr_freq_seg1_idx, false, false, &settings.freq_params);
 
 	if (channel) {
 		hostapd_get_op_class(candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].primary, candidate[iface->mxl_iface.sec_best_ch.second_rank_idx].width, &sec_op_class);
@@ -4621,6 +4632,54 @@ bool mxl_get_second_best_channel_from_acs_table(struct hostapd_iface *iface)
 		return false;
 	}
 }
+ 
+int mxl_acs_switch_to_unii_5_7_channel(struct hostapd_iface *iface)
+{
+	struct csa_settings settings;
+	struct hostapd_channel_data *channel = NULL;
+	int secondary_channel;
+	int can_idx;
+	u8 vht_oper_centr_freq_seg0_idx;
+	u8 vht_oper_centr_freq_seg1_idx;
+	struct acs_candidate_table *candidate = iface->mxl_iface.candidates;
+
+	/* set this flag true to get the idx value*/
+	iface->mxl_iface.sec_best_ch.flag = true;
+
+	os_memset(&settings, 0, sizeof(settings));
+	channel = mxl_dfs_get_channel_dynamic_failsafe(iface, &secondary_channel, &vht_oper_centr_freq_seg0_idx,
+					&vht_oper_centr_freq_seg1_idx, false, true, &settings.freq_params);
+	iface->mxl_iface.sec_best_ch.flag = false;
+	if (channel) {
+		can_idx = iface->mxl_iface.sec_best_ch.second_rank_idx;
+		iface->conf->channel = candidate[can_idx].primary;
+		iface->freq = hostapd_hw_get_freq(iface->bss[0], iface->conf->channel);
+		iface->conf->secondary_channel = mxl_get_secondary_offset(candidate[can_idx].primary, candidate[can_idx].secondary);
+		hostapd_set_oper_chwidth(iface->conf, mxl_get_oper_width(candidate[can_idx].width));
+		acs_smart_adjust_center_freq(iface, candidate[can_idx].chan);
+		wpa_printf(MSG_DEBUG, "ACS: UNII_5_7 channel idx %d,  chan : %d, freq : %d", can_idx, candidate[can_idx].primary, iface->freq);
+
+		/* Set channel and switch */
+		mxl_acs_chandef cur_chan;
+		acs_push_chandef(iface, &cur_chan);
+
+		mxl_hostapd_channel_switch_reason_set(iface, HAPD_CHAN_SWITCH_OTHER);
+
+		if (mxl_acs_do_switch_channel(iface, 0) < 0) {
+			/* back out with channel switch */
+			acs_pop_chandef(iface, &cur_chan);
+			wpa_printf(MSG_ERROR, "ACS Unable to switch to UNII_5_7 channel");
+			return -1;
+		}
+	}
+	else {
+		wpa_printf(MSG_ERROR, "ACS failed to get UNII_5_7 channel");
+   	return -2;
+	}
+
+	wpa_printf(MSG_ERROR, "ACS switch to UNII_5_7 channel successfull");
+	return 0;
+}
 
 void mxl_smart_acs_scan_complete(struct hostapd_iface *iface)
 {
diff --git a/src/ap/mxl_acs_proprietary.h b/src/ap/mxl_acs_proprietary.h
index bde829af2..43d115db1 100644
--- a/src/ap/mxl_acs_proprietary.h
+++ b/src/ap/mxl_acs_proprietary.h
@@ -198,7 +198,7 @@ int mxl_hostapd_get_chan_data_from_candidates_5g(int primary_chan, enum oper_cha
 int mxl_hostapd_acs_get_candidates_rank_idx(struct hostapd_iface *iface, int primary_channel,
 					    int secondary_offset, int bandwidth);
 bool mxl_hostapd_acs_compare_ranks(struct hostapd_iface *iface, int rank1_idx, int rank2_idx);
-struct acs_candidate_table *mxl_acs_get_failsafe_channel(struct hostapd_iface *iface, bool skip_usable);
+struct acs_candidate_table *mxl_acs_get_failsafe_channel(struct hostapd_iface *iface, bool skip_usable, bool only_unii_5_7);
 int mxl_hostapd_change_country(struct hostapd_iface *iface, struct channel_list_changed *info);
 int mxl_acs_find_center_chan(int freq, int width);
 void mxl_acs_bgscan_timer_cb(void *eloop_data, void *user_ctx);
diff --git a/src/ap/mxl_afc.c b/src/ap/mxl_afc.c
index abe00ca24..1e53fa880 100644
--- a/src/ap/mxl_afc.c
+++ b/src/ap/mxl_afc.c
@@ -16,7 +16,9 @@
 #include "utils/includes.h"
 #include "common/ieee802_11_defs.h"
 #include "hostapd.h"
+#include "hw_features.h"
 #include "mxl_afc.h"
+#include "mxl_acs.h"
 #include "ap_drv_ops.h"
 
 #define EIRP_UNIT_CONVERSION	100
@@ -305,6 +307,7 @@ int mxl_hostapd_afc_handle_request(struct hostapd_iface *iface)
 		 */
 		wpa_printf(MSG_ERROR, "AFC connection failed: %d", ret);
 		iface->afc.timeout = HOSTAPD_AFC_RETRY_TIMEOUT;
+		iface->afc.retry_count++;
 		if (ret == -EIO)
 			ret = -1;
 		goto resched;
@@ -316,12 +319,67 @@ int mxl_hostapd_afc_handle_request(struct hostapd_iface *iface)
 	     iface->afc.num_chan_info  == 0) ) {
 		wpa_printf(MSG_ERROR, "AFC has no valid rules");
 		iface->afc.timeout = HOSTAPD_AFC_RETRY_TIMEOUT;
+		iface->afc.retry_count++;
 		ret = -1;
 		goto resched;
 	}
+	else {
+		iface->afc.retry_count = 0;
+		iface->conf->he_6ghz_reg_pwr_type = iface->afc.he_6ghz_reg_pwr_type;
+	}
 
 resched:
-	mxl_hostapd_afc_resched_timeout(iface, iface->afc.timeout);
+	if (iface->conf->he_6ghz_reg_pwr_type == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR) {
+		mxl_hostapd_afc_resched_lpi(iface, HOSTAPD_AFC_RETRY_LPI);
+	}	
+	else {
+		mxl_hostapd_afc_resched_timeout(iface, iface->afc.timeout);
+	}
 
 	return ret;
 }
+
+int mxl_hostapd_afc_get_valid_rule_and_set_power(struct hostapd_iface *iface, int val)
+{
+	int ret = 0;
+
+	if (!is_6ghz_op_class(iface->conf->op_class) || !is_6ghz_freq(iface->freq))
+		return -1;
+
+	if (val == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR_SP) {
+		if(!(mxl_is_6ghz_unii5_unii7(iface->freq))) {
+#ifdef CONFIG_VENDOR_MXL_PROPRIETARY
+			ret = mxl_acs_switch_to_unii_5_7_channel(iface);
+			if(ret == -1) {
+				wpa_printf(MSG_ERROR, "AFC Failed to switch to UNII 5 or 7 channel for 6G composite power mode");
+				return -1;
+			}
+			else if (ret == -2) {
+				//TODO: select a fallback channel of UNII5 or UNII7
+				wpa_printf(MSG_ERROR, "AFC Failed to fetch UNII 5 or 7 channel, composite power mode not set");
+				return -1;
+			}
+#endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+		}
+	}
+	iface->afc.acs_pwr_mode_6g = 0;
+	iface->afc.he_6ghz_reg_pwr_type = val;
+	if (iface->afc.data_valid) {
+		iface->conf->he_6ghz_reg_pwr_type = val;
+		ret = mxl_hostapd_interface_update_power_regd(iface->bss[0], val);
+	}
+	else {
+		ret = mxl_hostapd_afc_handle_request(iface);
+		if (ret < 0) {
+			wpa_printf(MSG_ERROR, "AFC request failed retrying in few seconds");
+			return -1;
+		}
+		/* AFC request successfull trigger ACS and Switch channel */
+		hostapd_set_state(iface, HAPD_IFACE_ACS);
+		iface->afc.acs_pwr_mode_6g = 1;
+		/* Update power */
+		mxl_hostapd_interface_update_power_regd(iface->bss[0], val);
+	}
+
+ 	return ret;
+}
diff --git a/src/ap/mxl_afc.h b/src/ap/mxl_afc.h
index e3790a059..47d925d22 100644
--- a/src/ap/mxl_afc.h
+++ b/src/ap/mxl_afc.h
@@ -18,6 +18,7 @@
 #include "drivers/nl80211_copy.h"
 
 #define HOSTAPD_AFC_RETRY_TIMEOUT	180
+#define HOSTAPD_AFC_RETRY_LPI       12
 
 /* TODO: Move into vendor_cmds_copy.h after driver fix */
 struct ieee80211_freq_range {
@@ -81,5 +82,6 @@ int mxl_hostad_afc_update_chan_info(struct afc_chan_info_elem **chan_list,
 				    int *chan_list_size, u8 op_class, int center_chan, int power);
 int mxl_hostapd_afc_handle_request(struct hostapd_iface *iface);
 void mxl_hostapd_afc_resched_timeout(void *eloop_data, unsigned int secs);
-
+void mxl_hostapd_afc_resched_lpi(void *eloop_data, unsigned int secs);
+int mxl_hostapd_afc_get_valid_rule_and_set_power(struct hostapd_iface *iface, int val);
 #endif /* MXL_AFC_H */
diff --git a/src/ap/mxl_common_proprietary.c b/src/ap/mxl_common_proprietary.c
index aa5ed9b6c..a6fa3ce08 100644
--- a/src/ap/mxl_common_proprietary.c
+++ b/src/ap/mxl_common_proprietary.c
@@ -112,14 +112,14 @@ mxl_dfs_get_channel_dynamic_failsafe(struct hostapd_iface *iface,
                                      int *secondary_channel,
                                      u8 *vht_oper_centr_freq_seg0_idx,
                                      u8 *vht_oper_centr_freq_seg1_idx,
-                                     bool skip_usable,
+                                     bool skip_usable, bool only_unii_5_7,
                                      struct hostapd_freq_params* freq_params)
 {
 #ifdef CONFIG_ACS
 	int primary_freq;
 	struct hostapd_channel_data *channel;
 	struct acs_candidate_table *failsafe;
-	failsafe = mxl_acs_get_failsafe_channel(iface, skip_usable);
+	failsafe = mxl_acs_get_failsafe_channel(iface, skip_usable, only_unii_5_7);
 	int cf1;
 	struct mxl_hostapd_config *mxl_conf = &iface->conf->mxl_conf;
 
diff --git a/src/ap/mxl_dfs_common.c b/src/ap/mxl_dfs_common.c
index 586ef2f20..df524b10f 100644
--- a/src/ap/mxl_dfs_common.c
+++ b/src/ap/mxl_dfs_common.c
@@ -185,7 +185,7 @@ mxl_dfs_get_valid_channel_failsafe(struct hostapd_iface *iface,
 							       secondary_channel,
 							       vht_oper_centr_freq_seg0_idx,
 							       vht_oper_centr_freq_seg1_idx,
-							       skip_usable,
+							       skip_usable, false,
 							       NULL);
 	}
 
diff --git a/src/ap/mxl_dfs_common.h b/src/ap/mxl_dfs_common.h
index 714eb7481..3d400be55 100644
--- a/src/ap/mxl_dfs_common.h
+++ b/src/ap/mxl_dfs_common.h
@@ -81,7 +81,7 @@ int mxl_is_dfs_channel(int primary_freq_mhz, enum oper_chan_width chwidth);
 struct hostapd_channel_data *
 mxl_dfs_get_channel_dynamic_failsafe(struct hostapd_iface *iface, int *secondary_channel,
 				     u8 *vht_oper_centr_freq_seg0_idx, u8 *vht_oper_centr_freq_seg1_idx,
-				     bool skip_usable, struct hostapd_freq_params* freq_params);
+				     bool skip_usable, bool only_unii_5_7, struct hostapd_freq_params* freq_params);
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 void mxl_dfs_cut_cw_by_half(struct hostapd_iface *iface, bool backup);
 void mxl_dfs_restore_original_cw(struct hostapd_iface *iface);
diff --git a/src/ap/mxl_events.c b/src/ap/mxl_events.c
index 8dbd2bc7a..2891d78ab 100644
--- a/src/ap/mxl_events.c
+++ b/src/ap/mxl_events.c
@@ -199,7 +199,17 @@ static int mxl_hostapd_event_6ghz_regdb_update_info(struct hostapd_data *hapd, u
 		   mxl_he_6ghz_pwr_mode_text(hapd->iface->conf->he_6ghz_reg_pwr_type),
 		   mxl_he_6ghz_pwr_mode_text(hapd->iface->mxl_iface.he_6ghz_reg_pwr_mode));
 
-	if (he_reg_is_sp(hapd->iface->conf->he_6ghz_reg_pwr_type))
+	// Prevent switching back to SP mode if fallback to LPI mode is active
+#ifdef CONFIG_AFC
+   if (hapd->iface->mxl_iface.he_6ghz_reg_pwr_mode == HE_REG_INFO_6GHZ_AP_TYPE_INDOOR &&
+		flags == AFC_UPDATE_STATUS_DEFAULT && hapd->iface->afc.retry_count > 5) {
+		wpa_printf(MSG_INFO, "AFC fallback to LPI mode is active, ignoring Reg Update event");
+		hapd->iface->conf->he_6ghz_reg_pwr_type = HE_REG_INFO_6GHZ_AP_TYPE_INDOOR;
+		return 0;
+	}
+#endif /* CONFIG_AFC */
+
+	if (he_reg_is_sp(hapd->iface->conf->he_6ghz_reg_pwr_type) && !(he_reg_is_sp(hapd->iface->mxl_iface.he_6ghz_reg_pwr_mode)))
 		mxl_hostapd_interface_update_power_regd(hapd, hapd->iface->conf->he_6ghz_reg_pwr_type);
 
 	return 0;
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index af62d1eb3..ef72a2528 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2892,6 +2892,10 @@ int mxl_hostapd_no_ir_channel_list_updated(struct hostapd_iface *iface)
 			iface->conf->channel = 0;
 		}
 	}
+	if (iface->state == HAPD_IFACE_ACS) {
+		iface->freq = 0;
+		iface->conf->channel = 0;
+	}
 	mxl_setup_interface2(iface);
 	return 0;
 }
diff --git a/src/ap/mxl_zwdfs.c b/src/ap/mxl_zwdfs.c
index 295951d88..de3fc18da 100644
--- a/src/ap/mxl_zwdfs.c
+++ b/src/ap/mxl_zwdfs.c
@@ -716,7 +716,7 @@ void mxl_hostapd_zwdfs_cac_on_radar(struct hostapd_iface *iface, struct csa_sett
 						       &secondary_channel,
 						       &vht_oper_centr_freq_seg0_idx,
 						       &vht_oper_centr_freq_seg1_idx,
-						       true,
+						       true, false,
 						       &settings->freq_params);
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 
-- 
2.43.0

